### ISSUE TITLE
OpenSeadragon moving images unexpectedly with angles over 90 degreees

### ISSUE THREAD

**Issue by peterreeves** [openseadragon/openseadragon#1574]
I'm using OpenSeadragon in a project where we are trying to align images with each other. The images have arbitrary rotations, and I've noticed a problem whenever we have images over 90 degrees. The docs state that any angle outside the 0 to 90 range is mapped back into that, but is there a reason for this? When I try to align two images (which involves finding a translation, rotation, and scale) where one of them is rotated more than 90 degrees, the image shifts unexpectedly. The amount it shifts seems to be related to the aspect ratio of the image (square images don't seem to exhibit this behaviour). Unfortunately I can't seem to easily recreate the behavior, it only seems to happen on a few samples in the site I'm creating (All my artificial samples that I create don't seem to trigger it).

Is there a way to access the position more directly in OpenSeadragon? I know it rotates around the centre of the image, but can I get it to rotate around the top left corner always? It would simplify the maths a lot for me, and I think it would fix this issue. Otherwise, is there a formula or something that I can use to counteract this shift when rotating by more than 90 degrees?

**Comment by iangilman** [comment]
@avandecreme @nelsonn3c Any thoughts?

@peterreeves Thank you for reporting this! I'm actually not sure the best way to proceed, but isolating what causes it does seem like the next step; I'm sorry to hear you're having trouble doing that. As for where this is happening, you might look in TiledImage.drawTiles (for calls to getRotation). 

Are you doing anything else that might complicate the drawing, like using opacity or clipping or drawing with PNG tiles?

One thing to try (to isolate the issue) might be to set up an animation that slowly rotates through all of the rotations to see if anything pops out.

**Comment by nelsonn3c** [comment]
hum... It's awkward. When possible, please try to better reproduce your study case @peterreeves

**Comment by peterreeves** [comment]
Sorry for the slow response on this.

No, I'm not doing anything strange like using opacity or PNG tiles. Though surely those wouldn't affect the location on the screen?

I tried a whole bunch of fixes over the last month. I ended up just swapping the width and height lengths around in my own alignment code if the angle goes over 90 degrees, and this seems to have fixed it (or at least reduced the magnitude of the error).
This seems to be related to the way that OpenSeadragon does a similar thing with it's Rects?
https://openseadragon.github.io/docs/openseadragon.js.html#line14837

I don't fully understand what is going on though, I don't feel that this swapping of dimensions is correct (It feels like a massive hack).

In the future, it might be nice if OpenSeadragon could take more advanced input. It would be good if I could pass it a 3x3 or 4x4 matrix so I have precise control over how the image is drawn on screen (Most of my code works with matrices, I have to extract a position, rotation, and scale at the end to pass to OpenSeadragon). That would also allow me to skew and non-uniformly strech images which I have found the occasional need to do.

Consider this issue closed. If I end up digging a little deeper and finding a more satisfying answer I'll post it here.
